WP 6.1 deactivate Gutenberg plugin older than version 14.1

hellofromtonya opened the pull request:
## Details

Using the same strategy during WP 5.9 (see changeset https://core.trac.wordpress.org/changeset/52199):

* Renames the update function for WP 6.1
* Changes the Gutenberg version number from `11.9` to `14.1`

Why? Gutenberg plugin versions less than 14.1.0 are not compatible with WP 6.1, especially GB 13.9.0 to 14.0.x versions which will cause a fatal error.

## How to Test
Testing instructions are found here in the Trac ticket https://core.trac.wordpress.org/ticket/56985#comment:10

Trac ticket: https://core.trac.wordpress.org/ticket/56985

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Rename function for 6.1.1

hellofromtonya committed:
Change to GB 14.1

hellofromtonya committed:
Make inline more clear about version

hellofromtonya commented:
I've tested this PR. See the [Test Report](https://core.trac.wordpress.org/ticket/56985#comment:11).

Results:

- Gutenberg v 13.9.0: plugin deactivates during the update ✅ 
- Gutenberg v 14.0.3: plugin deactivates during the update ✅ 
- Gutenberg v 14.1.0: plugin stays activated (does not deactivate) during the update ✅ 

This PR fixes the issue.

desrosj reviewed (changes requested):
I'm not crazy about renaming the function. Even though it's private, the practice used in the areas of Core related to upgrades is to leave things in place for a historical and chronological reference. But I also recognize that duplicating all of this logic in a new `_upgrade_611_force_deactivate_incompatible_plugins()`.

What if we created a new variable similar to `$_new_bundled_files` and `$_old_files` used to map WP versions to minimum compatible version of a certain plugin, extract the deactivation logic into a new function (maybe `_upgrade_deactivate_incompatible_plugin`), and then just loop through?

Something like 
```
$_incompatible_plugin_versions = array(
	'gutenberg/gutenberg.php' => array(
		'5.9' => '11.9',
		'6.1' => '14.1',
	),
);
```

hellofromtonya commented:
@desrosj The renaming follows what happened in WP 5.9 (see https://core.trac.wordpress.org/changeset/52165 and https://core.trac.wordpress.org/changeset/52199).

When upgrading to the current version in `trunk` or the `branch`, the older versions no longer matter. Why? This change is not being backported to the older branches.

hellofromtonya commented:
The renaming and removal of the older function was discussed in the 5.9 ticket. Note a comment from @SergeyBiryukov https://core.trac.wordpress.org/ticket/54405#comment:16

>This is an interesting precedent as we don't generally remove functions, but it seems fine in this case, and makes sense to me too.

desrosj commented:
Ah, yes,. I see that now.

Sorry, I edited my comment above right after sending to include the following:

> Even though it's private, the practice used in the areas of Core related to upgrades is to leave things in place for a historical and chronological reference.

I think this would still be great to have, but since it's not currently the practice for this specific part, then this is _probably_ OK.

What about a compromise of removing the version from the function name and using the docblock (maybe @since tags) to track this history?
```
* @since 5.9.0 The minimum compatible version of Gutenberg is now 11.9.
* @since 6.1.0 The minimum compatible version of Gutenberg is now 14.1.
etc..
```

This makes it easier in the future (only the version number needs to be changed), and it also preserves a historical record.

hellofromtonya commented:
@desrosj Some questions:

* Are you also wanting to loop in the code from `_upgrade_440_force_deactivate_incompatible_plugins()` into this new `_upgrade_deactivate_incompatible_plugins()` function?
* Should we also remove `_upgrade_440_force_deactivate_incompatible_plugins()` and `_upgrade_590_force_deactivate_incompatible_plugins()` with this fix?

desrosj commented:
Maybe. I think that deserves a larger discussion though. We've demonstrated that though rare, these are not one off exceptions now (REST API, Gutenberg, etc.), so I think it makes sense to think about how to make this easier overall.

hellofromtonya commented:
@desrosj We wrote at the same time 🤣 Disregard my last comment. 

Maybe we could look at revisiting reworking this area of code in the next major cycle to incorporate your idea of modeling after the `$_new_bundled_files` and `$_old_files`. In doing so, this fix for this specific fatal error doesn't introduce any new risks and avoids scope creep.

hellofromtonya committed:
Remove versions. Add wp prefix

hellofromtonya committed:
Document version changes

hellofromtonya commented on src/wp-admin/includes/update-core.php:
```diff
@@ -1637,14 +1637,16 @@ function _upgrade_440_force_deactivate_incompatible_plugins() {
- * @since 5.9.0
+ * @since 5.8.0 The minimum compatible version of Gutenberg is 10.7.
+ * @since 5.9.0 The minimum compatible version of Gutenberg is 11.9.
+ * @since 6.1.1 The minimum compatible version of Gutenberg is 14.1.
  */
-function _upgrade_590_force_deactivate_incompatible_plugins() {
-	if ( defined( 'GUTENBERG_VERSION' ) && version_compare( GUTENBERG_VERSION, '11.9', '<' ) ) {
+function _wp_upgrade_force_deactivate_incompatible_plugins() {
```
@desrosj What do you think? 

* Added the `@since` for each release.
* Renamed the function to no longer be named by its release number.
* But added a `wp` prefix to avoid any naming conflicts outside of Core.

hellofromtonya replied:
Hmm, or maybe the function name should be:

```
_upgrade_core_force_deactivate_incompatible_plugins()
```

as it better aligns with the naming conventions in the same file. Thoughts?

hellofromtonya replied:
I went ahead and renamed in https://github.com/WordPress/wordpress-develop/pull/3559/commits/3588cef15aea3a36c4fc0dbb9b175dda3bae0f4c. The function name aligns better with the functions in the same file.

hellofromtonya committed:
Rename function for consistency

hellofromtonya commented on src/wp-admin/includes/update-core.php:
```diff
@@ -1637,14 +1637,16 @@ function _upgrade_440_force_deactivate_incompatible_plugins() {
- * @since 5.9.0
+ * @since 5.8.0 The minimum compatible version of Gutenberg is 10.7.
+ * @since 5.9.0 The minimum compatible version of Gutenberg is 11.9.
+ * @since 6.1.1 The minimum compatible version of Gutenberg is 14.1.
  */
-function _upgrade_590_force_deactivate_incompatible_plugins() {
-	if ( defined( 'GUTENBERG_VERSION' ) && version_compare( GUTENBERG_VERSION, '11.9', '<' ) ) {
+function _upgrade_core_deactivate_incompatible_plugins() {
```
@desrosj What do you think?

* Added the `@since` for each release. to document what changed.
* Renamed the function:
   * Removed the version number.
   * Added `core` to avoid naming conflicts outside of Core.
   * Removed the `force` for readability.
   
BTW Great idea on these changes 👏  You're right. Much better!

desrosj replied:
Looks great!

Only thing that is missing is the original `@since` tag noting when the function was introduced. But much better overall.

ironprogrammer replied:
It was originally `@since 5.8.0` (via c264ba80df1e400d5b1f19fe1a3fda96c7983a54), but the proposed change includes the added GB version detail 👍🏻

hellofromtonya replied:
I removed the comment from `@since 5.8.0` for consistency.

hellofromtonya replied:
@desrosj I'll assume with the final change made and your comment above, that it's okay to commit this PR. I'll mark it for `commit` and do the commit to `trunk`.

azaozz commented:
The patch LGTM too, good job! 

As discussed before, ideally this functionality will be in the plugins. It is not that hard to add a condition (or two) in a plugin's bootstrap to check the current WP version and stop loading the rest of the plugin (and show a warning, etc.) if it is deemed unsupported (too new). 

Of course some plugins would not want to do that as they are not updated frequently, and limiting the "lifetime" of current plugin version may not be desirable for the authors. However Gutenberg and other frequently updated plugins would be quite safer if they had such code.

desrosj commented:
> Of course some plugins would not want to do that as they are not updated frequently, and limiting the "lifetime" of current plugin version may not be desirable for the authors. However Gutenberg and other frequently updated plugins would be quite safer if they had such code.

I don't think it's that straightforward. In order for this to work, plugins need to know in advance which versions will be incompatible because even if the code is in a plugin release, there is no guarantee that site owners update that plugin. 

It's far more likely Core will update now that auto-updates for major releases are enabled by default. I think it's reasonable to have a small handful of exceptions for plugins built by Core team members and contributors. With the request for more canonical and feature plugins to build out and test new features, it's likely this problem will be encountered again soon.

azaozz commented:
> In order for this to work, plugins need to know in advance which versions will be incompatible because even if the code is in a plugin release, there is no guarantee that site owners update that plugin.

Right. Such cutout would be more of an estimate. It also sets a hard "Must update for compatibility with WP version x.y.z" requirement. So basically implementing a "Compatible up to" functionality would mean each plugin version will have a fixed lifecycle length.

I'd think that would be very welcome for Gutenberg as the code from the plugin is added to core regularly. Other frequently updated plugins would probably find it useful too. 

> I think it's reasonable to have a small handful of exceptions for plugins built by Core team members and contributors.

I'm not against having this function in core. It is a useful safeguard, and renamed as in this patch it looks pretty good. Just thinking it would be better for plugins to "manage" their compatibility themselves.

hellofromtonya committed:
Remove since 5.8.0 comment

5.8.0 is when the code was introduced. Removed the comment from the original version for consistency.

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/54789. Thank you everyone for your contributions!
